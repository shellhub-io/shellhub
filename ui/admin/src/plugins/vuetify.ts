// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
