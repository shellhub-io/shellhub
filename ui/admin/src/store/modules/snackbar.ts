import { Module } from "vuex";
import { State } from "./../index";

export interface snackbarMessageAndContentType {
  typeMessage: string;
  typeContent: string;
}

export interface SnackbarState {
  snackbarError: boolean;
  snackbarSuccess: boolean;
  snackbarMessageAndContentType: snackbarMessageAndContentType;
  snackbarCopy: boolean;
}

export const snackbar: Module<SnackbarState, State> = {
  namespaced: true,

  state: {
    snackbarError: false,
    snackbarSuccess: false,
    snackbarMessageAndContentType: { typeMessage: "", typeContent: "" },
    snackbarCopy: false,
  },

  getters: {
    snackbarSuccess: (state) => state.snackbarSuccess,
    snackbarError: (state) => state.snackbarError,
    snackbarMessageAndContentType: (state) => state.snackbarMessageAndContentType,
    snackbarCopy: (state) => state.snackbarCopy,
  },

  mutations: {
    setSnackbarSuccessAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarNoContent: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarSuccessDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarSuccess = true;
    },

    unsetSnackbarSuccess: (state) => {
      state.snackbarSuccess = false;
    },

    setSnackbarErrorLoadingOrAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorCustom: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarError = true;
    },

    setSnackbarErrorLicense: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    unsetSnackbarError: (state) => {
      state.snackbarError = false;
    },

    setSnackbarCopy: (state, value) => {
      state.snackbarMessageAndContentType = { typeMessage: "", typeContent: value };
      state.snackbarCopy = true;
    },

    unsetSnackbarCopy: (state) => {
      state.snackbarCopy = false;
    },
  },

  actions: {
    showSnackbarSuccessAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarSuccessAction", data);
    },

    showSnackbarSuccessDefault: ({ commit }) => {
      commit("setSnackbarSuccessDefault");
    },

    unsetShowStatusSnackbarSuccess: ({ commit }) => {
      commit("unsetSnackbarSuccess");
    },

    showSnackbarErrorLoading: ({ commit }, value) => {
      const data = { typeMessage: "loading", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarNoContent: ({ commit }) => {
      const data = { typeMessage: "no-content", typeContent: "" };
      commit("setSnackbarNoContent", data);
    },

    showSnackbarErrorAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorCustom: ({ commit }, value) => {
      const data = { typeMessage: "custom", typeContent: value };
      commit("setSnackbarErrorCustom", data);
    },

    showSnackbarErrorLicense: ({ commit }, value) => {
      const data = { typeMessage: "licenseRequired", typeContent: value };
      commit("setSnackbarErrorLicense", data);
    },

    showSnackbarErrorDefault: ({ commit }) => {
      commit("setSnackbarErrorDefault");
    },

    unsetShowStatusSnackbarError: ({ commit }) => {
      commit("unsetSnackbarError");
    },

    showSnackbarCopy: ({ commit }, value) => {
      commit("setSnackbarCopy", value);
    },

    unsetShowStatusSnackbarCopy: ({ commit }) => {
      commit("unsetSnackbarCopy");
    },
  },
};
