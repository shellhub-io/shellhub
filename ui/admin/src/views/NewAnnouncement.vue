<template>
  <h1>Create new Announcement</h1>

  <v-card class="mt-2">
    <v-card-item>
      <v-card-title>Title</v-card-title>

      <v-text-field
        v-model="title"
        :error-messages="titleError"
        variant="underlined"
        placeholder="Enter announcement title"
        data-test="announcement-title"
      />
    </v-card-item>

    <v-card-item class="mt-n4">
      <v-card-title>Content</v-card-title>

      <Editor
        :api-key="tinyMceKey"
        v-model="announcement"
        output-format="html"
        :init="{
          plugins: 'lists link image code help wordcount',
          menubar: 'file edit insert view tools help',
        }"
        toolbar="undo redo | blocks | fontsize image | bold italic link blockquote |
          \ bullist numlist | removeformat | help"
        data-test="announcement-content"
      />

      <v-alert
        v-if="announcementError"
        type="error"
        class="mt-2"
        data-test="announcement-error"
      >
        The announcement cannot be empty !
      </v-alert>

      <v-alert
        v-if="tinyMceKeyIsEmpty"
        type="warning"
        class="mt-2"
        data-test="announcement-key-warning"
      >
        It's recommended to set the TinyMCE key in the .env file.
      </v-alert>
    </v-card-item>

    <v-card-actions class="pa-4">
      <v-spacer />
      <v-btn
        @click="postAnnouncement"
        color="dark"
        variant="text"
        tabindex="0"
        data-test="announcement-btn-post"
      >
        Post
      </v-btn>
    </v-card-actions>
  </v-card>
</template>

<script setup lang="ts">
import { computed, ref, watch } from "vue";
import Editor from "@tinymce/tinymce-vue";
import TurndownService from "turndown";
import { useField } from "vee-validate";
import * as yup from "yup";
import { useRouter } from "vue-router";
import useSnackbarStore from "@admin/store/modules/snackbar";
import useAnnouncementStore from "@admin/store/modules/announcement";
import {
  INotificationsError,
  INotificationsSuccess,
} from "../interfaces/INotifications";
import { envVariables } from "../envVariables";

const router = useRouter();
const snackbarStore = useSnackbarStore();
const announcementStore = useAnnouncementStore();
const { value: title, errorMessage: titleError, setErrors: setTitleError } = useField<
      string | undefined
    >("title", yup.string().required(), {
      initialValue: "",
    });
const tinyMceKey = computed(() => envVariables.tinyMceKey);
const tinyMceKeyIsEmpty = computed(() => tinyMceKey.value === "");
const announcement = ref("");
const announcementError = ref(false);
const turndownService = new TurndownService();

watch(announcement, (val) => {
  if (val) announcementError.value = false;
});

const postAnnouncement = () => {
  if (!title.value) {
    setTitleError("Title cannot be empty !");
    return;
  }

  if (titleError.value || !announcement.value) {
    announcementError.value = true;
    snackbarStore.showSnackbarErrorAction(INotificationsError.announcementCreate);
    return;
  }

  try {
    const contentInHtml = turndownService.turndown(announcement.value);
    announcementStore.postAnnouncement({
      title: title.value,
      content: contentInHtml,
    });
    snackbarStore.showSnackbarSuccessAction(INotificationsSuccess.announcementCreate);
    router.push({ name: "announcements" });
  } catch (error) {
    snackbarStore.showSnackbarErrorAction(INotificationsError.announcementCreate);
  }
};

defineExpose({ title, titleError });
</script>

<style lang="scss">
.tox .tox-notification--warn, .tox .tox-notification--warning {
  display: none !important;
}
</style>
