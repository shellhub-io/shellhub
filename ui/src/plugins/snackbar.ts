import { App, Plugin, reactive } from "vue";

const SnackbarInjectionKey = Symbol("snackbar");

type SnackbarType = "success" | "error" | "info" | "warning";

interface SnackbarState {
  message: string;
  type: SnackbarType;
  show: boolean;
}

interface ISnackbarPlugin {
  showSuccess: (msg: string) => void;
  showError: (msg: string) => void;
  showInfo: (msg: string) => void;
  showWarning: (msg: string) => void;
  getMessage: () => string;
  getType: () => SnackbarType;
  getShow: () => boolean;
}

const state = reactive<SnackbarState>({
  message: "",
  type: "info",
  show: false,
});

const showSnackbar = (type: SnackbarType, message: string) => {
  state.message = message;
  state.type = type;
  state.show = true;

  setTimeout(() => {
    state.show = false;
  }, 4000);
};

const plugin: ISnackbarPlugin = {
  showSuccess: (msg: string) => showSnackbar("success", msg),
  showError: (msg: string) => showSnackbar("error", msg),
  showInfo: (msg: string) => showSnackbar("info", msg),
  showWarning: (msg: string) => showSnackbar("warning", msg),
  getMessage: () => state.message,
  getType: () => state.type,
  getShow: () => state.show,
};

const SnackbarPlugin: Plugin = {
  install(app: App) {
    app.provide(SnackbarInjectionKey, plugin);
  },
};

export { SnackbarPlugin, SnackbarInjectionKey, plugin };
export type { SnackbarState, ISnackbarPlugin };
