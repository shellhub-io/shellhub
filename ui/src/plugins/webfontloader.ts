// webfontloader documentation: https://github.com/typekit/webfontloader
const loadFonts = async () => {
  const webFontLoader = await import("webfontloader");

  webFontLoader.load({
    google: {
      families: ["Roboto:100,300,400,500,700,900&display=swap"],
    },
  });
};

export default loadFonts;
