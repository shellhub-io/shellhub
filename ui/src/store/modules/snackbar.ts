import { Module } from "vuex";
import { State } from "..";

export interface snackbarMessageAndContentType {
  typeMessage: string;
  typeContent: string;
}

export interface SnackbarState {
  message: string,
  type: string,
  snackbarError: boolean;
  snackbarSuccess: boolean;
  snackbarMessageAndContentType: snackbarMessageAndContentType;
  snackbarCopy: boolean;
}

export const snackbar: Module<SnackbarState, State> = {
  namespaced: true,
  state: {
    message: "",
    type: "",
    snackbarError: false,
    snackbarSuccess: false,
    snackbarMessageAndContentType: {
      typeMessage: "",
      typeContent: "",
    },
    snackbarCopy: false,
  },

  getters: {
    snackbarSuccess: (state) => state.snackbarSuccess,
    snackbarError: (state) => state.snackbarError,
    snackbarMessageAndContentType: (state) => state.snackbarMessageAndContentType,
    snackbarCopy: (state) => state.snackbarCopy,
  },

  mutations: {
    showMessage: (state, data) => {
      state.type = data.type;
      state.message = data.message;
    },

    setSnackbarSuccessAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    unsetSnackbarSuccess: (state) => {
      state.snackbarSuccess = false;
    },

    setSnackbarSuccessNotRequest: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarErrorLoadingOrAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorDefault: (state) => {
      state.snackbarMessageAndContentType = {
        typeMessage: "default",
        typeContent: "",
      };
      state.snackbarError = true;
    },

    unsetSnackbarError: (state) => {
      state.snackbarError = false;
    },

    setSnackbarCopy: (state, value) => {
      state.snackbarMessageAndContentType = {
        typeMessage: "",
        typeContent: value,
      };
      state.snackbarCopy = true;
    },

    unsetSnackbarCopy: (state) => {
      state.snackbarCopy = false;
    },
  },

  actions: {
    showSnackbarSuccessAction: ({ commit }, value) => {
      const data = {
        typeMessage: "action",
        typeContent: value,
      };
      commit("setSnackbarSuccessAction", data);
    },

    unsetShowStatusSnackbarSuccess: ({ commit }) => {
      commit("unsetSnackbarSuccess");
    },

    showSnackbarSuccessNotRequest: (context, value) => {
      const data = {
        typeMessage: "notRequest",
        typeContent: value,
      };
      context.commit("setSnackbarSuccessNotRequest", data);
    },

    showSnackbarErrorLoading: ({ commit }, value) => {
      const data = {
        typeMessage: "loading",
        typeContent: value,
      };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorAction: ({ commit }, value) => {
      const data = {
        typeMessage: "action",
        typeContent: value,
      };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorDefault: ({ commit }) => {
      commit("setSnackbarErrorDefault");
    },

    unsetShowStatusSnackbarError: ({ commit }) => {
      commit("unsetSnackbarError");
    },

    showSnackbarCopy: ({ commit }, value) => {
      commit("setSnackbarCopy", value);
    },

    unsetShowStatusSnackbarCopy: ({ commit }) => {
      commit("unsetSnackbarCopy");
    },
  },
};
