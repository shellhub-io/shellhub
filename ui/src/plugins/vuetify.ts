// Styles
import "@mdi/font/css/materialdesignicons.css";
import "@fortawesome/fontawesome-free/css/all.css";
import "vuetify/styles";
import { VFileUpload, VFileUploadItem } from "vuetify/labs/VFileUpload";
import "../styles/_variables.scss";
import { aliases as faAliases, fa } from "vuetify/iconsets/fa";

// Vuetify
import { createVuetify } from "vuetify";

const light = {
  dark: false,
  colors: {
    primary: "#667acc",
    secondary: "#FFFFFF",
    background: "#F5F5F5",
    tabs: "#F5F5F5",
    foreground: "#F5F5F5",
    paymentForm: "#F5F5F5",
    terminal: "#0f1526",
    "v-theme-background": "#FFFFFF",
    "v-theme-surface": "#FFFFFF",
    "v-theme-card": "#FFFFFF",
  },
};

const dark = {
  dark: true,
  colors: {
    primary: "#667acc",
    secondary: "#1E2127",
    background: "#18191B",
    tabs: "#1E1E1E",
    foreground: "#1E1E1E",
    paymentForm: "#E0E0E0",
    terminal: "#0f1526",
    "v-theme-background": "#1E2127",
    "v-theme-surface": "#1E2127",
    "v-theme-card": "#22252B",
  },
};

export default createVuetify({
  components: {
    VFileUpload,
    VFileUploadItem,
  },
  theme: {
    defaultTheme: "dark",
    themes: {
      dark,
      light,
    },
  },
  defaults: {
    global: {

    },
    VTextField: {
      variant: "outlined",
      density: "comfortable",
    },
    VSelect: {
      variant: "outlined",
      density: "comfortable",
    },
    VTextarea: {
      variant: "outlined",
    },
  },
  icons: {
    faAliases,
    sets: {
      fa,
    },
  },
});
