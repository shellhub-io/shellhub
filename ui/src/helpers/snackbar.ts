import { inject } from "vue";
import { InjectionKey, type ISnackbarPlugin } from "@/plugins/snackbar";

const useSnackbar = (): ISnackbarPlugin => inject(InjectionKey) as ISnackbarPlugin;

export default useSnackbar;
