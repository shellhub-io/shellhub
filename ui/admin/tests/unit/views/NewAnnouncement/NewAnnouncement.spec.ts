import { createVuetify } from "vuetify";
import { mount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { createPinia, setActivePinia } from "pinia";
import useAnnouncementStore from "@admin/store/modules/announcement";
import { SnackbarPlugin } from "@/plugins/snackbar";
import routes from "../../../../src/router";
import NewAnnouncement from "../../../../src/views/NewAnnouncement.vue";

type NewAnnouncementWrapper = VueWrapper<InstanceType<typeof NewAnnouncement>>;

describe("New Announcement", () => {
  let wrapper: NewAnnouncementWrapper;

  beforeEach(() => {
    const pinia = createPinia();
    setActivePinia(pinia);

    const announcementStore = useAnnouncementStore();

    announcementStore.postAnnouncement = vi.fn();
    announcementStore.fetchAnnouncements = vi.fn();

    const vuetify = createVuetify();

    wrapper = mount(NewAnnouncement, {
      global: {
        plugins: [pinia, vuetify, routes, SnackbarPlugin],
      },
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper).toBeTruthy();
  });

  it("Renders the component", () => {
    const html = wrapper.html().replace(/id="tiny-vue_\d+"/g, 'id="tiny-vue_random"');
    expect(html).toMatchSnapshot();
  });

  it("Has the correct title", () => {
    expect(wrapper.find("h1").text()).toBe("Create new Announcement");
  });

  it("Renders the correct HTML", () => {
    expect(wrapper.find("[data-test='announcement-title']").exists()).toBeTruthy();
    expect(wrapper.find("[data-test='announcement-content']").exists()).toBeTruthy();
    expect(wrapper.find("[data-test='announcement-error']").exists()).toBeFalsy();
    expect(wrapper.find("[data-test='announcement-btn-post']").exists()).toBeTruthy();
  });

  it("Renders the error message when the Title is empty", async () => {
    await wrapper.find("[data-test='announcement-btn-post']").trigger("click");
    expect(wrapper.vm.titleError).toBeTruthy();
  });

  it("Renders the error message when the announcement is empty", async () => {
    wrapper.vm.title = "News ShellHub";
    await wrapper.find("[data-test='announcement-btn-post']").trigger("click");
    expect(wrapper.find("[data-test='announcement-error']").exists()).toBeTruthy();
  });
});
