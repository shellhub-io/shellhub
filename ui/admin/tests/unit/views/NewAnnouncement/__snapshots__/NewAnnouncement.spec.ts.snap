// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`New Announcement > Renders the component 1`] = `
"<h1>Create new Announcement</h1>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated mt-2">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-item">
    <!---->
    <div class="v-card-item__content">
      <!---->
      <!---->
      <div class="v-card-title">Title</div>
      <div class="v-input v-input--horizontal v-input--density-default v-theme--light v-locale--is-ltr v-text-field v-input--plain-underlined" data-test="announcement-title">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--no-label v-field--variant-underlined v-theme--light v-locale--is-ltr">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div class="v-field__field" data-no-activator="">
              <!---->
              <!---->
              <!----><input placeholder="Enter announcement title" size="1" type="text" id="input-v-0" aria-describedby="input-v-0-messages" class="v-field__input" value="">
              <!---->
            </div>
            <!---->
            <!---->
            <div class="v-field__outline">
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <div id="input-v-0-messages" class="v-input__details" role="alert" aria-live="polite">
          <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
    </div>
    <!---->
  </div>
  <div class="v-card-item mt-n4">
    <!---->
    <div class="v-card-item__content">
      <!---->
      <!---->
      <div class="v-card-title">Content</div><textarea id="tiny-vue_random" visibility="hidden" data-test="announcement-content"></textarea>
      <!--v-if-->
      <!--v-if-->
    </div>
    <!---->
  </div>
  <div class="v-card-actions pa-4">
    <div class="v-spacer"></div><button type="button" class="v-btn v-btn--slim v-theme--light text-dark v-btn--density-default v-btn--size-default v-btn--variant-text" tabindex="0" data-test="announcement-btn-post"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""> Post </span>
      <!---->
      <!---->
    </button>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
